File: assignment2/clustering.m
function [c, Z] = clustering(X, Z0, opts)

%--------------------------------------------------------------------------
% INPUT
%--------------------------------------------------------------------------
%
%   - X     (n x m) matrix containing the m different data points.
%   - Z0    (n x k) initial points for the clusters (will be estimated by 
%                   Z0 = rand(n,3) if left empty).
%   - opts  field of options
%               .iter       number of maximum iterations (default = 100).
%               .display    print total distances (default = true).
%
%--------------------------------------------------------------------------

[n, m] = size(X);

%--------------------------------------------------------------------------
% PROCESS INPUT AND OPTIONS
%--------------------------------------------------------------------------

if nargin <= 1
    k = 3;
    Z = randn(n, k);
    opts.iter = 100;
    opts.display = true;
elif nargin <= 2
    k = size(Z0, 2);
    Z = Z0;
    opts.iter = 100;
    opts.display = true;
else
    k = size(Z0, 2);
    Z = Z0;
    if ~isfield(opts, 'iter')
       opts.iter = 100;
    end
    if ~isfield(opts, 'display')
       opts.display = true;
    end
end

sX = repmat(X, k, 1);

%--------------------------------------------------------------------------
% MAIN LOOP
%--------------------------------------------------------------------------

for i = 1:opts.iter
    % calculate distances
    sZ = abs(sX - reshape(Z(:), [], 1)).^2;
    sZ = sqrt(sum(reshape(sZ, n, m * k)));
    % select minimum distances
    sZ = reshape(sZ, k, m);
    [mZ, c] = min(sZ);

    % calculate new means
    for j = 1:k
        ind = (c == j);
        Z(:, j) = mean(X(:, ind), 2);
    end
end

% output
if opts.display
    s1 = sum(sum(mZ.^2));
    s2 = s1 / (m * k);
    fprintf(1, 'Total sum of distances = %5.2f; Weighted sum of distances = %1.6e\n', s1, s2);
end

end

File: assignment2/clustering.py
import numpy as np
import matplotlib.pyplot as plt
import time

# Assuming clustering function from clustering.py is available
def clustering(X, Z0=None, opts={'iter': 100, 'display': True}):
    n, m = X.shape
    if Z0 is None:
        k = 3
        Z0 = np.random.randn(n, k)
    else:
        k = Z0.shape[1]
    Z = Z0

    opts_alg = {}
    opts_alg['iter'] = opts.get('iter', 100)
    opts_alg['display'] = opts.get('display', True)

    sX = np.tile(X, (k, 1))

    for i in range(opts_alg['iter']):
        sZ = np.abs(sX - np.tile(Z.flatten('F').reshape(n * k, 1, order='F'), (1, m))) ** 2
        sZ = np.sqrt(sum(sZ.reshape(n, m * k, order='F')))
        sZ = sZ.reshape(k, m, order='F')
        mZ = np.min(sZ, axis=0)
        c = np.argmin(sZ, axis=0)
        for j in range(k):
            ind = (c == j)
            Z[:, j] = np.mean(X[:, ind], axis=1)

    if opts_alg['display']:
        s1 = np.sum(np.sum(mZ ** 2))
        s2 = s1 / (m * k)
        print('Total sum of distances = {:5.2f}; Weighted sum of distances = {:1.6e}'.format(s1, s2))

    return c, Z

# Function to generate random data clouds A, B, C
def generate_data(p):
    A = np.random.normal(0, 1, (2, p))
    B = 4 + np.random.normal(0, 1, (2, p))
    C = np.array([[5], [2]]) + np.random.normal(0, 1, (2, p))
    return np.hstack((A, B, C))  # Combine A, B, C to form X

# Set up data sizes for different runs
p_values = [100 * i for i in range(1, 11)]  # p = 100, 200, ..., 1000
mean_times = []

# Run clustering for different values of p and measure time
for p in p_values:
    X = generate_data(p)
    Z0 = np.random.randn(2, 3)  # Initial points for 3 clusters

    start_time = time.time()
    _, Z_final = clustering(X, Z0, opts={'iter': 100, 'display': False})
    end_time = time.time()

    elapsed_time = end_time - start_time
    mean_times.append(elapsed_time)

# Plot mean CPU time per iteration
plt.figure(figsize=(10, 6))
plt.plot(p_values, mean_times, marker='o')
plt.xlabel('Data Size (p)')
plt.ylabel('Mean CPU Time (s)')
plt.title('Mean CPU Time per Iteration for Different Data Sizes')
plt.grid(True)
plt.show()

# Visualize clustering result for the largest p
X = generate_data(p_values[-1])  # Use the largest p
Z0 = np.random.randn(2, 3)  # Initial points for 3 clusters

# Run clustering
c, Z_final = clustering(X, Z0, opts={'iter': 100, 'display': False})

# Visualize the clusters
plt.figure(figsize=(10, 6))
colors = ['r', 'g', 'b']
for j in range(3):
    plt.scatter(X[0, c == j], X[1, c == j], c=colors[j], label=f'Cluster {j+1}', alpha=0.5)

# Plot the final cluster centers
plt.scatter(Z_final[0, :], Z_final[1, :], c='k', marker='x', s=100, label='Cluster Centers')

plt.xlabel('X1')
plt.ylabel('X2')
plt.title('Clustering Results for Largest p')
plt.legend()
plt.grid(True)
plt.show()